design-proposals/vm-rollout-strategy/vm-rollout-strategy.md

```diff
@@ -0,0 +1,205 @@
+# Overview
+
+Today changes to a VM object (i.e. changing VM.spec.template) get applied, but only take effect upon the next restart - we call these pending changes “staged”. This is a proposal to introduce a custer-level rollout strategy field to control how changes to the VM object will propagate to the VirtualMachineInstance object without interupting the workload.
+
+
+## Motivation
+
+In most cases, the user would prefer the changes to the VM object to eventually propagate to the VMI object, as long as the rollout does not interrupt the workload. This behavior would be similar to how the Deployments roll out ReplicaSets.
+
+
+## Goals
+
+-   Provide a way to eventually propagate changes from a running VM object to its corresponding VMI without interrupting the workload.
+    
+-   Maintain backwards compatibility for cluster owners that prefer keeping staged changes until next VM reboot.
+    
+
+## Non Goals
+
+-   Live uppdate for **all** fields of the Virtual Machine API. Only select fields can be propagated to the guest. 
+-   Make LiveUpdate the default. We defer this decision.
+-   Support upgrades of VMs that are using the `liveUpdateFeatures`.
+    
+
+## Definition of Users
+
+A VM owner - with (at a minimum) create, update, and patch access for VM objects within a namespace.
+
+
+## User Stories
+
+As a VM owner I want changes that I applied to the VM object to propagate to the VMI if possible.
+
+As a VM owner, I want to be notified about changes that require a restart of the VM.
+
+## Repos
+
+kubevirt/kubevirt
+
+
+# Design
+
+> **Note:** Below we are only discussing changes to fields that support live updates.
+                  References to `all` fields only refers to fields that can be live-updated. 
+
+## KubeVirt CR
+
+A new field called `vmRolloutStrategy` will be introduced in the KubeVirt CR. This field will control the behavior of the VM objects. Two possible values will be available:
+
+1. `LiveUpdate` means all changes to the VM object will be eventually applied to the VMI.
+    
+2. `Stage` would mean the current familiar behavior remains - all changes to the VM object will be staged and applied after a restart.
```

Comment:
I am unsure about the CPU and memory, but for network interfaces there are 3 options to apply changes:
- Restart
- Live/On-the-fly
- Migration

I have no seen the migration option mentioned in this proposal.
It does not play any role with CPU and memory?

Reply:
It's just two cases:
- Restart (interrupt workload)
- LiveUpdate (Do not interrupt workload.

Then, live migration and on the fly are startegies for liveupdate.

IOW: Two states are enough IMO, and live update will lead to LM or on-the-fly according to what is possible - it's an implementation detail.

The same is btw true for mem/cpu once in-place pod updates / mutatable pods will GA.

Reply:
+1 for keeping only two strategies. If an when Network introduces an in-place hotplugging for certain devices, an additional feature gate can be added.

Reply:
That's an interesting topic. 

At this point, the hot plug of CPU and memory is only done via live migration, since pods are immutable.
However, in the future, we are planning to eventually use the [In place pod update resources](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources) feature, instead or to accompany live migration.

However, how hot plug is performed can be decided on the backend, instead of having a tunable for this.

I wonder how can a user decide which of these options is better. Why not make this decision on the backend?

Reply:
I also favor the two options from the user spec API.
I do think the condition added to the status needs to specify what is needed exactly: A reset of the domain/guest or a live-migration. From the UX perspective, the user may need to take action either way.

If you intend to automate migration as a result of detecting something, then it is possible but it needs to be explicit defined in the design. I am not sure if this is what we really want at the moment, because it has other implications, including complexity, that we do not know if it is a needed or not.

Reply:
> If you intend to automate migration as a result of detecting something, then it is possible but it needs to be explicit defined in the design. I am not sure if this is what we really want at the moment, because it has other implications, including complexity, that we do not know if it is a needed or not.

@EdDev Changes to the VM's CPU and Memory will trigger a live migration via the workload updater controller. This is how it is implemented today.

Reply:
Ok, thanks for the clarification.
Networking hotplug is not automating this, and I think not all will really want to automate it.

As long as this is left to the implementation, I think we are good.
IMO it will help to express in the condition the need for migration for the implementations that do not automate it. But this can be added later if you do not see the benefit now.

Reply:
There are a few concerns:
1. We don't see how one resource could reliably plug in new resource, if a different source has pending changes. this is due to the pod templating mechanism. Imagine new sockets are pending (thus they are in the spec already), if network now adds a new nic, then how do we prevent that the new socket (or rather the new cpu requests) end up in the new pod?
2. Because it is delcarative, are we sure that the declarative nature will be understandable to the user - what is applied and what not?
3. assume that we've got staged changes, and now a resource like net is pluggin in a new nic, but this fails, how do we clearly communicate these errors

concerns like this brought the authors to be more conservative about this point.

Reply:
The thing is: We are running out of time.
These uncertainties are the reason why it was defined to not cover this point.

In order to unblock this whole ... disucssion .... I'd advocating for leaving this undefined.